Modules/_ctypes/_ctypes.c

```diff
@@ -459,20 +444,112 @@ static PyType_Spec structparam_spec = {
     .slots = structparam_slots,
 };
 
+/*
+  CType_Type - a base metaclass. Its instances (classes) have a StgInfo.
+  */
+
+static int
+CType_Type_traverse(PyObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(Py_TYPE(self));
+    return 0;
+}
+
+static int
+CType_Type_clear(PyObject *self)
+{
+    ctypes_state *st = GLOBAL_STATE();
+    StgInfo *info;
+    if (PyStgInfo_FromType(st, self, &info) < 0) {
+        return -1;
+    }
+    if (info) {
+        Py_CLEAR(info->proto);
+        Py_CLEAR(info->argtypes);
+        Py_CLEAR(info->converters);
+        Py_CLEAR(info->restype);
+        Py_CLEAR(info->checker);
+    }
+    return 0;
+}
+
+static void
+CType_Type_dealloc(PyObject *self)
+{
+    ctypes_state *st = GLOBAL_STATE();
+    StgInfo *info;
+    if (PyStgInfo_FromType(st, self, &info) < 0) {
+        // ignore exception
+    }
+    if (info) {
+        PyMem_Free(info->ffi_type_pointer.elements);
+        PyMem_Free(info->format);
+        PyMem_Free(info->shape);
+    }
+
+    PyTypeObject *tp = Py_TYPE(self);
+    (void)CType_Type_clear(self);
+    PyType_Type.tp_dealloc(self);
+    Py_DECREF(tp);
+}
+
+static PyObject *
+CType_Type_sizeof(PyObject *self) {
+    Py_ssize_t size = Py_TYPE(self)->tp_basicsize;
+    size += Py_TYPE(self)->tp_itemsize * Py_SIZE(self);
+
+    ctypes_state *st = GLOBAL_STATE();
+    StgInfo *info;
+    if (PyStgInfo_FromType(st, self, &info) < 0) {
+        return NULL;
+    }
+    if (info) {
+        if (info->format) {
+            size += strlen(info->format) + 1;
+        }
+        if (info->ffi_type_pointer.elements) {
+            size += (info->length + 1) * sizeof(ffi_type *);
+        }
+        size += info->ndim * sizeof(Py_ssize_t);
+    }
+
+    return PyLong_FromSsize_t(size);
+}
+
+static PyMethodDef ctype_methods[] = {
+    {"__sizeof__", _PyCFunction_CAST(CType_Type_sizeof),
+     METH_NOARGS, PyDoc_STR("Return memory consumption of the type object.")},
+    {0},
+};
+
+static PyType_Slot ctype_type_slots[] = {
+    {Py_tp_traverse, CType_Type_traverse},
+    {Py_tp_clear, CType_Type_clear},
```

Comment:
```suggestion
```
Previously, `PyCStgDict_clear()` was not a `tp_clear` function.  `CType_Type_clear()` also seems to work just for `CType_Type_dealloc()` (mixable).
https://github.com/python/cpython/blob/7cee276d551a41d9271daf2a6bcd7def55555973/Modules/_ctypes/stgdict.c#L161-L162

Reply:
Alternatively, it would be less confusing to integrate `CDataType_*()` into `CType_Type_*()`, keeping the `{Py_tp_clear, CType_Type_clear}` slot for subtypes to inherit, roughly like the following:

<details>

```diff
 static int
 CType_Type_traverse(PyObject *self, visitproc visit, void *arg)
 {
+    ctypes_state *st = GLOBAL_STATE();
+    StgInfo *info;
+    if (PyStgInfo_FromType(st, self, &info) < 0) {
+        return -1;
+    }
+    if (info) {
+        Py_VISIT(info->proto);    // GC-related
+        Py_VISIT(info->argtypes);
+        Py_VISIT(info->converters);
+        Py_VISIT(info->restype);
+        Py_VISIT(info->checker);
+    }
     Py_VISIT(Py_TYPE(self));
-    return 0;
+    return PyType_Type.tp_traverse(self, visit, arg);
 }

 static int
 CType_Type_clear(PyObject *self)
 {
     ctypes_state *st = GLOBAL_STATE();
     StgInfo *info;
     if (PyStgInfo_FromType(st, self, &info) < 0) {
         return -1;
     }
     if (info) {
         Py_CLEAR(info->proto);    // GC-related
         Py_CLEAR(info->argtypes);
         Py_CLEAR(info->converters);
         Py_CLEAR(info->restype);
         Py_CLEAR(info->checker);
     }
-    return 0;
+    return PyType_Type.tp_clear(self);
 }

 static PyType_Slot foo_type_slots[] = {
     ...
-    {Py_tp_traverse, CDataType_traverse},
-    {Py_tp_clear, CDataType_clear},
     ...
 };

 PyType_Spec foo_type_spec = {
     ...
-    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC |
+    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE |
     ...
 };
```
</details>

Reply:
That would indeed be clearer! I'd prefer leaving it to a follow-up PR though: it'll be easier to debug when we get to isolating the module state.